ext/standard/streamsfuncs.c

```diff
@@ -756,115 +795,144 @@ static int stream_array_emulate_read_fd_set(zval *stream_array)
-PHP_FUNCTION(stream_select)
-{
-	zval *r_array, *w_array, *e_array;
-	struct timeval tv, *tv_p = NULL;
-	fd_set rfds, wfds, efds;
-	php_socket_t max_fd = 0;
-	int retval, sets = 0;
-	zend_long sec, usec = 0;
-	bool secnull;
-	bool usecnull = 1;
-	int set_count, max_set_count = 0;
-
-	ZEND_PARSE_PARAMETERS_START(4, 5)
-		Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
-		Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
-		Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
-		Z_PARAM_LONG_OR_NULL(sec, secnull)
-		Z_PARAM_OPTIONAL
-		Z_PARAM_LONG_OR_NULL(usec, usecnull)
-	ZEND_PARSE_PARAMETERS_END();
+PHP_FUNCTION(stream_select) {
+    zval *r_array, *w_array, *e_array;
+    struct timeval tv, *tv_p = NULL;
+    fd_bigset rfds, wfds, efds;
+    php_socket_t max_fd = 0;
+    int retval, sets = 0;
+    zend_long sec, usec = 0;
+    bool secnull;
+    bool usecnull = 1;
+    int set_count = 0, max_set_count = 0;
+
+    ZEND_PARSE_PARAMETERS_START(4, 5)
+        Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
+        Z_PARAM_LONG_OR_NULL(sec, secnull)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_LONG_OR_NULL(usec, usecnull)
+    ZEND_PARSE_PARAMETERS_END();
 
-	FD_ZERO(&rfds);
-	FD_ZERO(&wfds);
-	FD_ZERO(&efds);
+#ifdef PHP_WIN32
+	/* Check if the Windows standard handle is redirected to file */
+	long max_fds = FD_SETSIZE;
+#else
+	/* Check if the file descriptor identifier is a terminal */
+	long max_fds = sysconf(_SC_OPEN_MAX);
+#endif
 
-	if (r_array != NULL) {
-		set_count = stream_array_to_fd_set(r_array, &rfds, &max_fd);
+    FD_BIGSET_ZERO(&rfds, max_fds);
+    FD_BIGSET_ZERO(&wfds, max_fds);
+    FD_BIGSET_ZERO(&efds, max_fds);
+
+    if (r_array != NULL) {
+        set_count = stream_array_to_fd_bigset(r_array, &rfds, &max_fd, max_fds);
 		if (set_count > max_set_count)
 			max_set_count = set_count;
-		sets += set_count;
+		sets += set_count;    
 	}
 
-	if (w_array != NULL) {
-		set_count = stream_array_to_fd_set(w_array, &wfds, &max_fd);
+    if (w_array != NULL) {
+        set_count = stream_array_to_fd_bigset(w_array, &wfds, &max_fd, max_fds);
 		if (set_count > max_set_count)
 			max_set_count = set_count;
 		sets += set_count;
 	}
 
-	if (e_array != NULL) {
-		set_count = stream_array_to_fd_set(e_array, &efds, &max_fd);
+    if (e_array != NULL) {
+        set_count = stream_array_to_fd_bigset(e_array, &efds, &max_fd, max_fds);
 		if (set_count > max_set_count)
 			max_set_count = set_count;
-		sets += set_count;
-	}
+        sets += set_count;
+    }
 
-	if (!sets) {
-		zend_value_error("No stream arrays were passed");
-		RETURN_THROWS();
-	}
+    if (!sets) {
+        zend_value_error("No stream arrays were passed");
+        FD_BIGSET_FREE(&rfds);
+        FD_BIGSET_FREE(&wfds);
+        FD_BIGSET_FREE(&efds);
+        RETURN_THROWS();
+    }
 
-	if (!PHP_SAFE_MAX_FD(max_fd, max_set_count)) {
+	if (max_set_count == 0) {
```

Comment:
Are we missing `FD_BIGSET_FREE` before returning?

Reply:
I'm not sure where you mean; I think FD_BIGSET_FREE is everywhere it is needed now, but there was a few places they were missing before the last push.

Reply:
We've already called `FD_BIGSET_ZERO` by this point, so the memory is allocated. On this line, we are about to return without freeing that memory; thus I think we need to free it.

That being said, I'm not quite sure how we would get here.